### DISCUSSION TITLE
Terrasonnet - Tanka-like implementation for Terraform

### DISCUSSION THREAD

**Discussion by ghostsquad** [grafana/tanka#589] [Ideas]
@sh0rez - I've been working with Jsonnet, generating terraform/hcl code (as json), and I was thinking of writing a tool, like Tanka. I've already come up with a name: `terrasonnet`. I was wondering if this is something with which you'd be interested in collaborating with me. I'd like it to have feature parity with terragrunt as well.

**Comment by Duologic** [comment]
Hey @ghostsquad 

Within Grafana Labs, we're exploring new ways to leverage Terraform too. Initially I would like to come up with a good way to render Jsonnet into JSON/HCL. 

I've experimented a little bit here: https://github.com/jsonnet-libs/k8s/blob/master/jsonnet/terraform.jsonnet 
Also @jdbaldry has experimented in this area with https://github.com/jdbaldry/tfsonnet

Do you have any experiments going perhaps?

**Reply by ghostsquad** [reply]
Internally at TuneIn I'm rendering providers, variables, locals, resources, modules, and outputs for a given environment for spinning up kubernetes clusters. Using a similar approach as terragrunt, in that I'm using a directory hierarchy to automatically set specific things (such as region, or account) for a cluster within that hierarchy.

I've leveraged jsonnet to give me overrides, and to provide easy mechanisms to add "mixins" (to allow me to use composition not just inheritance)

Jsonnet has allowed me to do things such as dynamically declared module sources (`source` natively can't contain interpolation) and other areas that were not possible or easy to do from hcl alone.

**Reply by Duologic** [reply]
That sounds really cool, I would like to see more of that. Is there anything you can share publicly?

**Reply by ghostsquad** [reply]
I'll put together an example and share it.

**Reply by tskinn** [reply]
@ghostsquad If you start a project I would be more than happy to try and contribute. I've been looking for something like this.

**Reply by ghostsquad** [reply]
Jsonnet, even Terraform continue to be problematic. I came to realize that jsonnet on top of HCL was just more problematic. Now you needed to know 2 languages, how they interact, and when to do something in jsonnet vs in HCL.

I've begun to look at Pulumi instead. If I'm forced to use Terraform, I may continue down this path. Otherwise, Pulumi, the Pulumi K8s Operator, etc look really promising.

Terrasonnet feels like a future ksonnet. A dead end.

I think jsonnet is still quite awesome though, and do wish that language can grow, but I'm skeptical of that too.

I'm still up to show some code though if folks are still interested.

**Reply by tskinn** [reply]
Yeah, after I wrote this I spent some time experimenting again and ran into some familiar problems. I think I agree with you. Dead end.

**Reply by Duologic** [reply]
I'm currently implementing Crossplane, it has the potential to solve a few of the problems we face with Terraform. We have a generated jsonnet library for creating Crossplane objects here: https://jsonnet-libs.github.io/crossplane-libsonnet/

**Comment by frimik** [comment]
I’m using terramate for all my dynamic and code generation needs for terraform now. 

The way I tie terraform Tanka (jsonnet) together now is I read from terraform variable files both from Terraform and in Jsonnet.

I use terraform.tfvars.json anywhere I want to also use the same data in Jsonnet. 

Then I have certain Terraform stacks which generate `/lib/terraform-data/…/config.json` files which also gets added and committed to the source repo which also can be read in by Jsonnet in the CI pipeline. This is not so much to strictly fully automate, but to primarily single source data and to eliminate human factors which otherwise may lead to copy-paste errors or configuration drift.

**Reply by ghostsquad** [reply]
This sounds very similar to what I've done. Personally I haven't had many issues with code generation, but folks I've worked with haven't been as receptive for a variety of reasons.

I wish Jsonnet was easier to debug. That seems like the biggest issue. The errors and stack traces are seldom helpful.

Seems years later, Jsonnet is still not advancing at any reasonable pace. I think I've stopped caring about solving problems like this. Developers don't want to trade a higher cognitive burden for the flexibility that Jsonnet and code generation provides.

**Comment by Duologic** [comment]
This project seems to be quite active on this topic: https://docs.tflibsonnet.com/
